java-checks/src/main/java/org/sonar/java/checks/SillyEqualsCheck.java

```diff
@@ -0,0 +1,124 @@
+    .name("equals")
+    .addParameter(JAVA_LANG_OBJECT);
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree methodTree = (MethodInvocationTree) tree;
+    if (EQUALS_METHOD.matches(methodTree)) {
+      ExpressionTree firstArgument = Iterables.getOnlyElement(methodTree.arguments());
+      Type argumentType = firstArgument.symbolType().erasure();
+      Type ownerType = getMethodOwnerType(methodTree).erasure();
+      if (isLiteralNull(firstArgument)) {
+        addIssue(tree, "Remove this call to \"equals\"; comparisons against null always return false; consider using '== null' to check for nullity.");
+      } else if (ownerType.isArray()) {
+        checkWhenOwnerIsArray(tree, ownerType, argumentType);
+      } else {
+        checkWhenOwnerIsNotArray(tree, ownerType, argumentType);
+      }
+    }
+  }
+
+  private void checkWhenOwnerIsArray(Tree tree, Type ownerType, Type argumentType) {
+    if (!argumentType.isArray()) {
+      if (!argumentType.is(JAVA_LANG_OBJECT)) {
+        addIssue(tree, "Remove this call to \"equals\"; comparisons between an array and a type always return false.");
+      }
+    } else if (!areRelated(((Type.ArrayType) ownerType).elementType(), ((Type.ArrayType) argumentType).elementType())) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons between unrelated arrays always return false.");
+    } else {
+      addIssue(tree, "Use \"Arrays.equals(array1, array2)\" or the \"==\" operator instead of using the \"Object.equals(Object obj)\" method.");
+    }
+  }
+
+  private void checkWhenOwnerIsNotArray(Tree tree, Type ownerType, Type argumentType) {
+    if (argumentType.isArray() && !ownerType.is(JAVA_LANG_OBJECT)) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons between a type and an array always return false.");
+    } else if (argumentType.isClass() && !areRelated(ownerType, argumentType)) {
+      if (isFinalClassWithInterface(ownerType, argumentType)) {
+        addIssue(tree, MESSAGE);
+      } else if (!ownerType.symbol().isInterface() && !argumentType.symbol().isInterface()) {
+        addIssue(tree, MESSAGE);
+      }
+    }
+  }
+
+  private boolean isFinalClassWithInterface(Type type1, Type type2) {
+    return (type1.symbol().isInterface() && type2.symbol().isFinal()) || (type2.symbol().isInterface() && type1.symbol().isFinal());
+  }
+
+  private boolean isLiteralNull(Tree tree) {
+    return tree.is(Tree.Kind.NULL_LITERAL);
+  }
+
+  private Type getMethodOwnerType(MethodInvocationTree methodSelectTree) {
+    if (methodSelectTree.methodSelect().is(Tree.Kind.MEMBER_SELECT)) {
+      return ((MemberSelectExpressionTree) methodSelectTree.methodSelect()).expression().symbolType();
+    } else {
+      return methodSelectTree.symbol().owner().type();
+    }
+  }
+
+  private boolean areRelated(Type type1, Type type2) {
```

Comment:
This method is called only negated, so it should probalby be renamed and include the negation.